stdlib/public/TensorFlow/ShapedArray.swift

```diff
@@ -264,14 +264,174 @@ internal extension _ShapedArrayProtocol {
+  /// left.
+  func leftPadded(toLength length: Int) -> String {
+    return repeatElement(" ", count: Swift.max(0, length - count)) + self
+  }
+}
+
+/// Common public protocol implementations.
+
 fileprivate extension _ShapedArrayProtocol
-  where Element : _ShapedArrayProtocol {
-  var _description: String {
+  where Element : _ShapedArrayProtocol, Element == Element.Element
+{
+  /// Returns the whitespace separator between elements, given the current
+  /// indent level.
+  func separator(indentLevel: Int) -> String {
+    if rank == 1 {
+      return ", "
+    }
+    return String(repeating: "\n", count: rank - 1) +
+           String(repeating: " ", count: indentLevel + 1)
+  }
+
+  /// A textual representation of the 1-D shaped array, starting at the given
+  /// indent level. Returns a summarized description if `summarize` is true and
+  /// the element count exceeds twice the `edgeElementCount`.
+  ///
+  /// - Parameters:
+  ///   - indentLevel: The indentation level.
+  ///   - edgeElementCount: The maximum of elements to print before and after
+  ///     summarization via ellipses (`...`).
+  ///   - maxScalarLength: The length of the longest scalar description in the
+  ///     entire original array-to-print.
+  ///   - scalarCountPerLine: The number of scalars to print per line, used when
+  ///     printing 1-D vectors.
+  ///   - summarize: If true, summarize description if element count exceeds
+  ///     twice `edgeElementCount`.
+  func vectorDescription(
+    indentLevel: Int, edgeElementCount: Int, maxScalarLength: Int,
+    scalarCountPerLine: Int, summarize: Bool
+  ) -> String {
+    // Get scalar descriptions.
+    func scalarDescription(_ element: Element) -> String {
+      let description = String(describing: element)
+      return description.leftPadded(toLength: maxScalarLength)
+    }
+    var scalarDescriptions: [String] = []
+    if summarize && count > 2 * edgeElementCount {
+      scalarDescriptions +=
+        prefix(edgeElementCount).map(scalarDescription)
+      scalarDescriptions += ["..."]
+      scalarDescriptions +=
+        suffix(edgeElementCount).map(scalarDescription)
+    } else {
+      scalarDescriptions += map(scalarDescription)
+    }
+
+    // Combine scalar descriptions into lines, based on the scalar count per
+    // line.
+    let lines = stride(
+      from: 0, to: scalarDescriptions.count, by: scalarCountPerLine
+    ).map { i -> ArraySlice<String> in
+      let upperBound = Swift.min(
+        i.advanced(by: scalarCountPerLine), scalarDescriptions.count)
+      return scalarDescriptions[i..<upperBound]
+    }
+
+    // Return lines joined with separators.
+    let lineSeparator = ",\n" + String(repeating: " ", count: indentLevel + 1)
+    return lines.enumerated().reduce(into: "[") { result, entry in
+      let (i, line) = entry
+      result += line.joined(separator: ", ")
+      result += i != lines.count - 1 ? lineSeparator : ""
+    } + "]"
```

Comment:
Why do a concatenation with `"]"` at the end, when, i think, you can make the string literal at the end of line 339 be `"]"`?

Reply:
Not quite - with your alternative, when `lines` is empty, the result is `"["`.

We discussed in person, and concluded that much code could be made more efficient using iterators/a single for-loop. I'll defer that for now to save it - improving performance here would be a nice Swift starter bug.